File: go800mon/a800mon/trainer.go
package a800mon

import "errors"

type Trainer struct {
	startAddr  uint16
	endAddr    uint16
	length     int
	value      byte
	hasValue   bool
	snapshot   []byte
	candidates []int
	reader     func(start uint16, length int) ([]byte, error)
}

type TrainerRow struct {
	Addr  uint16
	Value byte
}

func NewTrainer(start uint16, end uint16, value *byte) (*Trainer, error) {
	if end < start {
		return nil, errors.New("Stop address must be >= start address.")
	}
	t := &Trainer{
		startAddr: start,
		endAddr:   end,
		length:    int(end-start) + 1,
	}
	if value != nil {
		t.value = *value
		t.hasValue = true
	}
	return t, nil
}

func (t *Trainer) BindReader(reader func(start uint16, length int) ([]byte, error)) {
	t.reader = reader
}

func (t *Trainer) Start(value *byte) (int, error) {
	if value != nil {
		t.value = *value
		t.hasValue = true
	}
	if !t.hasValue {
		return 0, errors.New("Missing trainer start value.")
	}
	current, err := t.read()
	if err != nil {
		return 0, err
	}
	t.snapshot = current
	t.candidates = t.candidates[:0]
	target := t.value
	for idx, b := range current {
		if b == target {
			t.candidates = append(t.candidates, idx)
		}
	}
	return len(t.candidates), nil
}

func (t *Trainer) Changed(value byte) (int, error) {
	current, err := t.read()
	if err != nil {
		return 0, err
	}
	next := make([]int, 0, len(t.candidates))
	for _, idx := range t.candidates {
		if current[idx] != t.snapshot[idx] && current[idx] == value {
			next = append(next, idx)
		}
	}
	t.candidates = next
	t.snapshot = current
	t.value = value
	t.hasValue = true
	return len(t.candidates), nil
}

func (t *Trainer) NotChanged() (int, error) {
	current, err := t.read()
	if err != nil {
		return 0, err
	}
	next := make([]int, 0, len(t.candidates))
	for _, idx := range t.candidates {
		if current[idx] == t.snapshot[idx] {
			next = append(next, idx)
		}
	}
	t.candidates = next
	t.snapshot = current
	return len(t.candidates), nil
}

func (t *Trainer) Reset() {
	t.snapshot = nil
	t.candidates = nil
}

func (t *Trainer) MatchCount() int {
	return len(t.candidates)
}

func (t *Trainer) Rows(limit int) []TrainerRow {
	total := len(t.candidates)
	if limit <= 0 || limit > total {
		limit = total
	}
	rows := make([]TrainerRow, 0, limit)
	for _, idx := range t.candidates[:limit] {
		rows = append(rows, TrainerRow{
			Addr:  uint16((int(t.startAddr) + idx) & 0xFFFF),
			Value: t.snapshot[idx],
		})
	}
	return rows
}

func (t *Trainer) read() ([]byte, error) {
	if t.reader == nil {
		return nil, errors.New("Trainer reader is not bound.")
	}
	data, err := t.reader(t.startAddr, t.length)
	if err != nil {
		return nil, err
	}
	if len(data) < t.length {
		return nil, errors.New("Trainer read returned too few bytes.")
	}
	return append([]byte(nil), data[:t.length]...), nil
}

File: py800mon/trainer.py
class Trainer:
    def __init__(self, start: int, end: int, value: int | None = None):
        self.start_addr = start & 0xFFFF
        self.end_addr = end & 0xFFFF
        self.length = (self.end_addr - self.start_addr) + 1
        self.value = None if value is None else (value & 0xFF)
        self.snapshot = b""
        self.candidates = []
        self._reader = None

    def bind_reader(self, reader):
        self._reader = reader

    def start(self, value: int | None = None):
        if value is not None:
            self.value = value & 0xFF
        if self.value is None:
            raise ValueError("Missing trainer start value.")
        current = self._read()
        self.snapshot = current
        target = self.value
        self.candidates = [idx for idx, b in enumerate(current) if b == target]
        return len(self.candidates)

    def changed(self, value: int):
        current = self._read()
        target = value & 0xFF
        self.candidates = [
            idx
            for idx in self.candidates
            if current[idx] != self.snapshot[idx] and current[idx] == target
        ]
        self.snapshot = current
        self.value = target
        return len(self.candidates)

    def not_changed(self):
        current = self._read()
        self.candidates = [
            idx for idx in self.candidates if current[idx] == self.snapshot[idx]
        ]
        self.snapshot = current
        return len(self.candidates)

    def reset(self):
        self.snapshot = b""
        self.candidates = []

    def rows(self, limit: int | None = None):
        total = len(self.candidates)
        shown = total if limit is None or limit > total else limit
        return [
            ((self.start_addr + offset) & 0xFFFF, self.snapshot[offset])
            for offset in self.candidates[:shown]
        ]

    def _read(self):
        if not self._reader:
            raise RuntimeError("Trainer reader is not bound.")
        return self._reader(self.start_addr, self.length)
